operator/internal/controller/pulumi/stack_controller.go

```diff
@@ -654,7 +632,16 @@ func (r *StackReconciler) Reconcile(ctx context.Context, request ctrl.Request) (
 	}
 
 	// Step 2. If there are extra environment variables, read them in now and use them for subsequent commands.
-
+	err = sess.setupWorkspace(ctx)
+	if err != nil {
+		if isStalledError(err) {
+			// TODO: why assume it is a cross-namespace error?
+			instance.Status.MarkStalledCondition(pulumiv1.StalledCrossNamespaceRefForbiddenReason, err.Error())
+			return reconcile.Result{}, saveStatus()
+		}
+		log.Error(err, "Failed to setup Pulumi workspace")
+		return reconcile.Result{}, err
+	}
```

Comment:
I decoupled the generic `setupWorkspace` from the source-specific code, for testability.  i.e. `SetupWorkspaceFromFluxSource` doesn't call `setupWorkspace`  anymore.